src/ngraph/op/batch_norm.cpp

```diff
@@ -269,13 +271,31 @@ void ngraph::op::BatchNorm::generate_adjoints(autodiff::Adjoints& adjoints,
     std::vector<std::shared_ptr<Node>> goes(get_outputs().size());
     if (this->get_training_flag() && get_input_size() == 3)
     {
+        std::set<size_t> output_input_found;
```

Comment:
@varun-intel There needs to be a comment describing what this is all about.
@Krovatkin Whatever this is doing, it's way too complicated if it needs to be done this way.

Reply:
this is due to a fact that users aren't ordered but this code needs to be able to tell which between output, mean or variance, so we need to recover that order. I'm going to add a simple helper that will hopefully make it less painful.... but there isn't a very good way around this problem.

Reply:
@diyessi @varun-intel with #1154 
we could do something like this : 
```cpp

//TODO: define get_goe_or_default to create a new GOE if one doesn't exist already
//We could define `get_goe_or_default` in `BatchNorm` class, then we wouldn't need the second argument.

auto goes = op::GetOutputElement::get_get_output_elements(this->shared_from_this());
auto output = get_goe_or_default(goes.at(0), this->shared_from_this());
auto mean = get_goe_or_default(goes.at(1), this->shared_from_this());
auto var = get_goe_or_default(goes.at(2), this->shared_from_this());

```